yaf-src/airframe/doc/airdaemon.1
.\" Automatically generated by Pod::Man 2.27 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "AIRDAEMON 1"
.TH AIRDAEMON 1 "14-Oct-2021" "2.12.1" "airframe"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
airdaemon \- Run a program as a daemon process, restarting it if it dies
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 5
\&    airdaemon   [\-\-retry RETRY_DELAY] [\-\-retry\-max RETRY_MAX_DELAY]
\&                [\-\-pidfile PID_FILE] [\-\-airdaemon\-pidfile PID_FILE]
\&                [\-\-log LOG_SPECIFIER] [\-\-loglevel LOG_LEVEL]
\&                [\-\-verbose] [\-\-version] [\-\-no\-daemon]
\&                \-\- PROGRAM [PROGRAM_ARGS]
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBairdaemon\fR invokes a child process and ensures that it is restarted if it
encounters any errors.  The delay between child exit and restart can be
chosen, and can either be constant or exponentially increasing up to a
specified maximum.
.PP
Use of two dashes (\-\-) after all airdaemon command-line switches allows
\&\s-1PROGRAM_ARGS\s0 to be interpreted by the \s-1PROGRAM\s0 rather than \fBairdaemon\fR itself.
While they are not strictly required if you do not need to pass arguments to
\&\s-1PROGRAM,\s0 they should be used for consistency.
.SH "OPTIONS"
.IX Header "OPTIONS"
.SS "Daemon Options"
.IX Subsection "Daemon Options"
.IP "\fB\-\-retry\fR=\fI\s-1RETRY_DELAY\s0\fR" 4
.IX Item "--retry=RETRY_DELAY"
Wait \s-1RETRY_DELAY\s0 seconds after the child process exits abnormally until trying
to restart the child process.  See also \fB\-\-retry\-max\fR
.IP "\fB\-\-retry\-max\fR=\fI\s-1RETRY_MAX_DELAY\s0\fR" 4
.IX Item "--retry-max=RETRY_MAX_DELAY"
This parameter, combined with the \fB\-\-retry\fR parameter, facilitates truncated
binary exponential backoff of the retry logic.  The first retry attempt will
be delayed by the value of the \fB\-\-retry\fR parameter.  If the child process
exits abnormally within this time period, the delay is doubled.  Once the
child process survives longer than the value of the \fB\-\-retry\fR parameter, the
retry delay is reset to that value.
.Sp
For example, assume \fB\-\-retry\fR=30 and \fB\-\-retry\-max\fR=240.  The first time the
child process exits abnormally, \fBairdaemon\fR will wait 30 seconds before
invoking the child process again.  If it exits abnormally again within 30
seconds, \fBairdaemon\fR will wait 60 seconds and try again.  If the process
dies again within 30 seconds, the delay doubles again to 120.  The process
continues until the delay hits the maximum of 240.  Once the child process
survives for 30 seconds, the delay will be reset to the original retry delay
of 30.
.IP "\fB\-\-pidfile\fR=\fI\s-1PID_FILE\s0\fR" 4
.IX Item "--pidfile=PID_FILE"
Write the process identifier of the child process to \fI\s-1PID_FILE\s0\fR.
Each time the child process is restarted, the process identifier in this file
is updated. This option exists to facilitate the termination of the child
process by shutdown scripts. Note that if the child process terminates
normally during shutdown, airdaemon will terminate normally as well, so it
is usually sufficient just to kill the pid in this file at shutdown.
.IP "\fB\-\-airdaemon\-pidfile\fR=\fI\s-1PID_FILE\s0\fR" 4
.IX Item "--airdaemon-pidfile=PID_FILE"
Write the process identifier of the airdaemon process to \fI\s-1PID_FILE\s0\fR.
This option exists to facilitate the termination of the forked
airdaemon process by shutdown scripts.
.IP "\fB\-\-no\-daemon\fR" 4
.IX Item "--no-daemon"
Do not actually daemonize.  Mainly useful for testing/debugging.
.SS "Logging Options"
.IX Subsection "Logging Options"
These options are used to specify how log messages are routed.  airdaemon can
log to standard error, regular files, or the \s-1UNIX\s0 syslog facility.
.IP "\fB\-\-log\fR \fI\s-1LOG_SPECIFIER\s0\fR" 4
.IX Item "--log LOG_SPECIFIER"
Specifies destination for log messages. \fI\s-1LOG_SPECIFIER\s0\fR can be a \fIsyslog\fR\|(3)
facility name, the special value \fBstderr\fR for standard error, or the
\&\fIabsolute\fR path to a file for file logging. Standard error logging is only
available in \fB\-\-daemon\fR mode if \fB\-\-foreground\fR is present. The default log
specifier is \fBstderr\fR if available, \fBuser\fR otherwise.
.IP "\fB\-\-loglevel\fR \fI\s-1LOG_LEVEL\s0\fR" 4
.IX Item "--loglevel LOG_LEVEL"
Specify minimum level for logged messages. In increasing levels of verbosity,
the supported log levels are \fBquiet\fR, \fBerror\fR, \fBcritical\fR, \fBwarning\fR,
\&\fBmessage\fR, \fBinfo\fR, and \fBdebug\fR. The default logging level is \fBwarning\fR.
.IP "\fB\-\-verbose\fR" 4
.IX Item "--verbose"
Equivalent to \fB\-\-loglevel debug\fR.
.IP "\fB\-\-version\fR" 4
.IX Item "--version"
If present, print version and copyright information to standard error and exit.
.SH "EXAMPLES"
.IX Header "EXAMPLES"
The following will ping www.cert.org every 15 seconds until ping returns
sucess:
.PP
.Vb 1
\&    airdaemon \-\-retry 15 \-\- ping \-c 1 www.cert.org
.Ve
.PP
The same as above, except the delay will double until reaching a maximum of 15
minutes:
.PP
.Vb 1
\&    airdaemon \-\-retry 15 \-\-retry\-max 900 \-\- ping \-c 1 www.cert.org
.Ve
.SH "BUGS"
.IX Header "BUGS"
Known issues are listed in the \fB\s-1README\s0\fR file in the Airframe source
distribution. Note that airdaemon should be considered alpha-quality software;
not every concievable input and aggregation is exhaustively tested at each
release, and specific features may be completely untested. Please be mindful
of this before deploying airdaemon in production environments. Bug reports and
feature requests may be sent via email to <netsa\-help@cert.org>.
.SH "AUTHORS"
.IX Header "AUTHORS"
Tony Cebzanov and  Brian Trammell for the
\&\s-1CERT\s0 Network Situational Awareness
Group, http://www.cert.org/netsa
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIfiledaemon\fR\|(1)


yaf-src/NEWS
Version 2.12.2: 2021-10-14
==========================

Added new protocols to the yafAppLabelRules.conf file and updated several regular expressions.

Changed the regexes used by the SMTP DPI plugin and improved capture when multiple messages appear in a single SMTP session.

Fixed a crash in the SMTP DPI plugin when reading uniflow records.

Updated the POP3 DPI plugin.

Updated yafzcbalance to be compatibile with PF_Ring-8.

Version 2.12.1: 2020-12-22
==========================

Changed the templates and IEs used for SMTP DPI.  The new templates use different IDs than those used by previous releases of YAF.  super_mediator-1.8.0 or later is required to read this format.  Currently there is no version of Analysis Pipeline that reads the SMTP DPI.

First public release of YAF 2.12.x.

Version 2.11.3: 2021-10-14
==========================

Added new protocols to the yafAppLabelRules.conf file and updated several regular expressions.

Updated the SMTP regexes used in the yafDPIRules.conf file to be similar to those used by the smtpplugin in YAF 2.12.2.

Updated yafzcbalance to be compatibile with PF_Ring-8.

Note: The yafAppLabelRules.conf file yafDPIRules.conf files in this release are usuable in prior releases of YAF also.

Version 2.11.2: 2020-11-20
==========================

Corrected the patch to allow building with PF_Ring.

Version 2.11.1: 2020-11-19
==========================

Fixed bugs in NTP and DNS deep packet inspection.

Fixed a compilation error when building with metadata export enabled.

Fixed possible compilation errors when building with nDPI support.

Fixed compilation errors when building with newer versions of PF_Ring.

Version 2.11.0: 2019-03-18
==========================

Support for FixBuf 2.3.0 added, and is now required.

Added support for nDPI 2.0.

CERT Info Model support added.

More strict DNS applabel.

Initial NTP Mode 7 applabel support.

Improved POSIX compliance for init script.

Removed ipfixDump; it is now distributed with libfixbuf.

DNS DPI free segfault fix.

New YAF stats and tombstone format.

Version 2.10.0: 2018-04-30
==========================

Support for FixBuf 2.0.0 added, and is now required.

Derive information elements from included XML files.

Various reporting/output bug fixes for ipfixDump.

Support for tombstone records added.

Version 2.9.3: 2017-12-21
==========================

Fixed configure-time dependency for libndpi to limit use of v1.8.0 and greater.

Modified init script to give YAF more time to shut down gracefully.

Version 2.9.2: 2017-11-08
==========================

Fixed configure-time bug when using libfixbuf 1.7.1 (or earlier) and p0fprinter

Version 2.9.1: 2017-11-02
==========================

Fixed bug that could corrupt flow emitted to standard output

Version 2.9.0: 2017-10-19
==========================

nDPI library support added

Added NTP applabel

Added RFC5610 template metadata (name and description) record output (libfixbuf 1.8.0 or greater required).

Add option --no-vlan-in-key to drop VLAN ID from hash calculation

Minor Bug Fixes

Version 2.8.4: 2016-04-14
=========================

Fix incompatibility with older versions of libpcap introduced in 2.8.3

Version 2.8.3: 2016-04-12
=========================

Important bug fix for versions 2.8.x.  Fixes a bug in decoding specific TCP Options headers.

Version 2.8.2: 2016-04-05
=========================

Fix application labeling bug introduced in 2.8.0 which incorrectly labels particular REGEX labels

Other Bug Fixes

Version 2.8.1: 2016-02-04
==========================

Fix compile error when configured with --disable-payload

Force buffer emit with IPFIX Options record when inactive 


Version 2.8.0: 2015-12-22
==========================

Remove support for fixbuf releases prior to libfixbuf-1.7.0

PF_RING support

PF_RING ZC (Zero Copy) support

Add support for gzip'd PCAP files

Add support for decoding MPTCP headers and exporting MPTCP information

Add LUA configuration file for yaf startup

New SSL Server Name field export from TLS/SSL Client Hello

New option for exporting entire X.509 Certificate

Add Fragment flag to flowAttributes to signify that a flow contained fragmented packets

DHCP fingerprinting plugin now exports basic list of options by default

ipfixDump prints number of records for each template

Bug Fix for labeling DNS over TCP

Bug Fix for reverseFlowDeltaMilliseconds field

Bug Fix for collecting X.509 Certificates through a proxy

More detailed information about ignored packets on termination/SIGUSR1

Version 2.7.1: 2015-01-27
=========================

Fix a bug with --flow-stats in particular configurations

Version 2.7.0: 2015-01-07
=======================

New Gh0st RAT Application Label

New NetBIOS Datagram Service Application Label

yafMeta2Pcap can now accept IPFIX input

getFlowKeyHash now exports IPFIX

Support for indexing PCAPNG files

New YAF option --no-output to produce no IPFIX output

New YAF options --hash and --stime to search for a single flow with the given hash and start time

DNS DPI now exports query section of resource record for all responses with nonzero RCODE

Faster searching of pcap-meta files

Implement SAME_SIZE flag for TCP flows

Minor Bug Fixes

Version 2.6.0: 2014-09-03
=================

Added a new tool, ipfixDump, to read and dump the contents of IPFIX files. Requires Fixbuf 1.4.0 or later.

Add LDAP application label

Filedaemon can now move files from one directory to another without passing to a child program

SSL/TLS DPI modification to capture SSL record version

Update CERT PEN Information Elements to use full information model if Fixbuf 1.4.0 or later is available

Fix for Modbus application label to reduce false positives

Bug Fix for TOS field when running with --uniflow

Bug Fix in RPM spec file

Bug Fix for labeling malformed DNS packets

Bug Fix for processing out of order packets with --force-read-all

Bug Fix for exporting reverse payload

Other minor bug fixes


yaf-src/airframe/src/airdaemon.c
/*
** airdaemon.c
** Keeps a child process running.
**
** ------------------------------------------------------------------------
** Copyright (C) 2007-2021 Carnegie Mellon University. All Rights Reserved.
** ------------------------------------------------------------------------
** Authors: Tony Cebzanov <toyc@cert.org>
** ------------------------------------------------------------------------
** @OPENSOURCE_HEADER_START@
** Use of the YAF system and related source code is subject to the terms
** of the following licenses:
**
** GNU General Public License (GPL) Rights pursuant to Version 2, June 1991
** Government Purpose License Rights (GPLR) pursuant to DFARS 252.227.7013
**
** NO WARRANTY
**
** ANY INFORMATION, MATERIALS, SERVICES, INTELLECTUAL PROPERTY OR OTHER
** PROPERTY OR RIGHTS GRANTED OR PROVIDED BY CARNEGIE MELLON UNIVERSITY
** PURSUANT TO THIS LICENSE (HEREINAFTER THE "DELIVERABLES") ARE ON AN
** "AS-IS" BASIS. CARNEGIE MELLON UNIVERSITY MAKES NO WARRANTIES OF ANY
** KIND, EITHER EXPRESS OR IMPLIED AS TO ANY MATTER INCLUDING, BUT NOT
** LIMITED TO, WARRANTY OF FITNESS FOR A PARTICULAR PURPOSE,
** MERCHANTABILITY, INFORMATIONAL CONTENT, NONINFRINGEMENT, OR ERROR-FREE
** OPERATION. CARNEGIE MELLON UNIVERSITY SHALL NOT BE LIABLE FOR INDIRECT,
** SPECIAL OR CONSEQUENTIAL DAMAGES, SUCH AS LOSS OF PROFITS OR INABILITY
** TO USE SAID INTELLECTUAL PROPERTY, UNDER THIS LICENSE, REGARDLESS OF
** WHETHER SUCH PARTY WAS AWARE OF THE POSSIBILITY OF SUCH DAMAGES.
** LICENSEE AGREES THAT IT WILL NOT MAKE ANY WARRANTY ON BEHALF OF
** CARNEGIE MELLON UNIVERSITY, EXPRESS OR IMPLIED, TO ANY PERSON
** CONCERNING THE APPLICATION OF OR THE RESULTS TO BE OBTAINED WITH THE
** DELIVERABLES UNDER THIS LICENSE.
**
** Licensee hereby agrees to defend, indemnify, and hold harmless Carnegie
** Mellon University, its trustees, officers, employees, and agents from
** all claims or demands made against them (and any related losses,
** expenses, or attorney's fees) arising out of, or relating to Licensee's
** and/or its sub licensees' negligent use or willful misuse of or
** negligent conduct or willful misconduct regarding the Software,
** facilities, or other rights or assistance granted by Carnegie Mellon
** University under this License, including, but not limited to, any
** claims of product liability, personal injury, death, damage to
** property, or violation of any laws or regulations.
**
** Carnegie Mellon University Software Engineering Institute authored
** documents are sponsored by the U.S. Department of Defense under
** Contract FA8721-05-C-0003. Carnegie Mellon University retains
** copyrights in all material produced under this contract. The U.S.
** Government retains a non-exclusive, royalty-free license to publish or
** reproduce these documents, or allow others to do so, for U.S.
** Government purposes only pursuant to the copyright license under the
** contract clause at 252.227.7013.
**
** @OPENSOURCE_HEADER_END@
** ------------------------------------------------------------------------
*/

#define _AIRFRAME_SOURCE_
#include <airframe/autoinc.h>
#include <airframe/airopt.h>
#include <airframe/logconfig.h>

uint32_t       ad_retry_min        = 30;
uint32_t       ad_retry_max        = 0;
gboolean       ad_nodaemon         = FALSE;
pid_t          ad_pid              = 0;
char          *ad_pidfile          = NULL;
pid_t          ad_cpid              = 0;
char          *ad_cpidfile          = NULL;


AirOptionEntry ad_options[]  = {
    AF_OPTION( "retry", 'r', 0, AF_OPT_TYPE_INT, &ad_retry_min,
               "Retry delay in seconds", "sec" ),
    AF_OPTION( "retry-max", 'R', 0, AF_OPT_TYPE_INT, &ad_retry_max,
               "Retry delay maximum in seconds", NULL ),
    AF_OPTION( "pidfile", 'P', 0, AF_OPT_TYPE_STRING, &ad_cpidfile,
               "A filename to write the child process pid to", NULL ),
    AF_OPTION( "airdaemon-pidfile", 'A', 0, AF_OPT_TYPE_STRING, &ad_pidfile,
               "A filename to write airdaemon's pid to", NULL ),
    AF_OPTION( "no-daemon", (char)0, 0, AF_OPT_TYPE_NONE, &ad_nodaemon,
               "do not daemonize", NULL ),
    AF_OPTION_END
};

typedef struct _ad_child_data {
    GMainLoop  *loop;
    gboolean   *done;
} ad_child_data_t;

static void
parse_options(
    int   *argc,
    char **argv[])
{
    AirOptionCtx *aoctx = NULL;

    aoctx = air_option_context_new("", argc, argv, ad_options);
    logc_add_option_group(aoctx, "airdaemon", VERSION);

    air_option_context_set_help_enabled(aoctx);

    air_option_context_parse(aoctx);
}


static void
on_child_exit(
    GPid       child_pid,
    gint       status,
    gpointer   data)
{
    GMainLoop *loop = ((ad_child_data_t *)data)->loop;
    gboolean  *done = ((ad_child_data_t *)data)->done;

    g_message("pid %lu exited with status %d", (gulong)child_pid, status);

#ifdef G_OS_UNIX
    if (WIFEXITED(status)) {
        if (WEXITSTATUS(status) == EXIT_SUCCESS) {
            g_debug("pid %lu returned success", (gulong)child_pid);
            *done = TRUE;
        } else {
            g_warning("pid %lu returned error status %d", (gulong)child_pid,
                      WEXITSTATUS(status));
        }
    } else if (WIFSIGNALED(status)) {
        g_critical("pid %lu terminated with signal %d\n",
                   (gulong)child_pid, WTERMSIG(status));
    } else {
        g_critical("pid %lu terminated", (gulong)child_pid);
    }
#endif /* G_OS_UNIX */
    g_spawn_close_pid(child_pid);
    g_main_loop_quit(loop);
}


gboolean
daemonize(
    void)
{
    /* fork */
    if (fork()) {exit(0);}

    /* dissociate from controlling terminal */
    if (setsid() < 0) {
        g_critical("setsid() failed: %s", strerror(errno));
        return FALSE;
    }

    /* redirect stdio */
    freopen("/dev/null", "r", stdin);
    freopen("/dev/null", "w", stdout);
    freopen("/dev/null", "w", stderr);

    ad_pid = getpid();
    if (ad_pidfile) {
        FILE *pidfile = fopen(ad_pidfile, "w");
        if (!pidfile) {
            g_critical("could not write pidfile");
            goto end;
        }
        fprintf(pidfile, "%d\n", ad_pid);
        fclose(pidfile);
    }
  end:

    return TRUE;
}


int
main(
    int    argc,
    char  *argv[])
{
    int        i;
    gboolean   done = FALSE;
    GError    *err = NULL;
    uint32_t   delay;
    GTimer    *uptimer  = NULL;
    gdouble    elapsed_time;

    GMainLoop *loop;

    GPtrArray *child_args     = NULL;

    /* parse options */
    parse_options(&argc, &argv);

    /* set up logging */
    if (!logc_setup(&err)) {
        air_opterr("%s", err->message);
    }

    if (ad_retry_max && (ad_retry_min > ad_retry_max) ) {
        air_opterr("--retry value (%d) cannot exceed --retry-max value (%d) ",
                   ad_retry_min,
                   ad_retry_max);
    }
    delay = ad_retry_min;

    child_args = g_ptr_array_sized_new(64);
    for (i = 1; i < argc; i++) {
        /* Double dash indicates end of airdaemon's arguments */
        if (!strncmp(argv[i], "--", strlen(argv[i])) ) {
            continue;
        }
        g_ptr_array_add(child_args, g_strdup(argv[i]));
    }
    g_ptr_array_add(child_args, NULL);

    loop = g_main_loop_new(NULL, FALSE);

    /* Options check out; daemonize */
    if (!ad_nodaemon) {
        if (!daemonize()) {
            goto end;
        }
    }

    uptimer = g_timer_new();

    while (!done) {
        GPid            child_pid;
        char          **child_envp            = {NULL};
        GError         *child_err             = NULL;
        ad_child_data_t child_data;

        if (!g_spawn_async_with_pipes(".",
                                      (gchar **)child_args->pdata,
                                      child_envp,
                                      G_SPAWN_SEARCH_PATH |
                                      G_SPAWN_DO_NOT_REAP_CHILD,
                                      NULL,
                                      NULL,
                                      &child_pid,
                                      NULL,
                                      NULL,
                                      NULL,
                                      &child_err))
        {
            g_error("error spawning process: %s",
                    (child_err && child_err->message ? child_err->
                     message : "unknown error"));
        }

        g_timer_start(uptimer);

        /* Write child pid if requested */
        if (ad_cpidfile) {
            FILE *cpidfile = fopen(ad_cpidfile, "w");
            if (!cpidfile) {
                g_critical("could not write pidfile");
                goto end;
            }
            fprintf(cpidfile, "%d\n", child_pid);
            fclose(cpidfile);
        }

        /* Watch for process exit status */
        child_data.loop = loop;
        child_data.done = &done;

        g_child_watch_add(child_pid, on_child_exit, &child_data);
        g_main_loop_run(loop);

        g_timer_stop(uptimer);
        elapsed_time = g_timer_elapsed(uptimer, NULL);

        if (done) {
            g_debug("done");
        } else {
            if (ad_retry_max && (elapsed_time >= ad_retry_min) ) {
                g_debug("child survived for %fs, resetting delay",
                        elapsed_time);
                delay = ad_retry_min;
            }
            g_debug("child exited abnormally, sleeping for %d seconds", delay);
            sleep(delay);
            if (ad_retry_max) {
                if (2 * delay <= ad_retry_max) {
                    delay *= 2;
                } else {
                    delay = ad_retry_max;
                }
            }
        }
    }

  end:

    return 0;
}
